wandb/sync/sync.py

```diff
@@ -70,20 +77,118 @@ def __init__(
+            elif TFEVENT_SUBSTRING in sync_item:
+                tb_root = os.path.dirname(os.path.abspath(sync_item))
+                tb_logdirs.append(tb_root)
+                tb_event_files = 1
+        return tb_event_files, tb_logdirs, tb_root
+
+    def _send_tensorboard(self, tb_root, tb_logdirs, sm, publish_interface, record_q):
+        if self._entity is None:
+            viewer, server_info = sm._api.viewer_server_info()
+            self._entity = viewer.get("entity")
+        proto_run = wandb_internal_pb2.RunRecord()
+        proto_run.run_id = self._run_id or wandb.util.generate_id()
+        proto_run.project = self._project or wandb.util.auto_project_name(None)
+        proto_run.entity = self._entity
+        url = "{}/{}/{}/runs/{}".format(
+            self._app_url,
+            url_quote(proto_run.entity),
+            url_quote(proto_run.project),
+            url_quote(proto_run.run_id),
+        )
+        print("Syncing: %s ..." % url)
+        sys.stdout.flush()
+        record = publish_interface._make_record(run=proto_run)
+        sm.send(record)
+        # TODO: temp directory not working?
+        settings = wandb.Settings(
+            root_dir=TMPDIR.name,
+            run_id=proto_run.run_id,
+            _start_datetime=datetime.datetime.now(),
+            _start_time=time.time(),
+        )
+        tb_watcher.TBHistory.ignore = lambda key: "bert" in key
+        watcher = tb_watcher.TBWatcher(settings, proto_run, publish_interface, True)
+        for tb in tb_logdirs:
+            watcher.add(tb, True, tb_root)
+            sys.stdout.flush()
+        watcher.finish()
+        # send all of our records like a boss
+        while not record_q.empty():
+            data = record_q.get(block=True)
+            sm.send(data)
+        sys.stdout.flush()
+        sm.finish()
 
     def run(self):
         for sync_item in self._sync_list:
+            tb_event_files, tb_logdirs, tb_root = self._find_tfevent_files(sync_item)
             if os.path.isdir(sync_item):
                 files = os.listdir(sync_item)
                 filtered_files = list(filter(lambda f: f.endswith(WANDB_SUFFIX), files))
-                if check_and_warn_old(files) or len(filtered_files) != 1:
+                if tb_root is None and (
+                    check_and_warn_old(files) or len(filtered_files) != 1
+                ):
                     print("Skipping directory: {}".format(sync_item))
                     continue
-                sync_item = os.path.join(sync_item, filtered_files[0])
+                if len(filtered_files) > 0:
+                    sync_item = os.path.join(sync_item, filtered_files[0])
             dirname = os.path.dirname(sync_item)
             files_dir = os.path.join(dirname, "files")
+            # If we're syncing tensorboard, let's use a tmpdir
+            if not os.path.exists(files_dir):
+                files_dir = os.path.join(TMPDIR.name, "files")
```

Comment:
At this point, we aren't quite sure if we are syncing tensorboard. It seems like the assumption is if the `sync_item/files` dir does not exist, we presume that we are in the tensorboard state? This might be fragile if for whatever reason the `files` dir was deleted from a true wandb dir, or if the user has a `files` subdirectory.

Reply:
Nice catch @tssweeney.  I actually thought about the missing case, but not the user dir case.  I'll fix.